pydantic_ai_slim/pydantic_ai/prompt_config.py

```diff
@@ -0,0 +1,414 @@
+from __future__ import annotations as _annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass, replace
+from textwrap import dedent
+from typing import TYPE_CHECKING, Any
+
+import pydantic_core
+
+from pydantic_ai.usage import RunUsage
+
+from ._run_context import RunContext
+from .messages import ModelMessage, ModelRequest, ModelRequestPart, RetryPromptPart, ToolReturnPart
+
+if TYPE_CHECKING:
+    from pydantic_ai.agent import Agent
+    from pydantic_ai.models import Model
+
+__all__ = (
+    # templates & configuration
+    'PromptConfig',
+    'PromptTemplates',
+    'ToolConfig',
+    # defaults
+    'DEFAULT_FINAL_RESULT_PROCESSED',
+    'DEFAULT_OUTPUT_TOOL_NOT_EXECUTED',
+    'DEFAULT_OUTPUT_VALIDATION_FAILED',
+    'DEFAULT_FUNCTION_TOOL_NOT_EXECUTED',
+    'DEFAULT_TOOL_CALL_DENIED',
+    'DEFAULT_MODEL_RETRY',
+    'DEFAULT_PROMPTED_OUTPUT_TEMPLATE',
+)
+
+# Default template strings - used when template field is None
+DEFAULT_FINAL_RESULT_PROCESSED = 'Final result processed.'
+"""Default confirmation message when a final result is successfully processed."""
+
+DEFAULT_OUTPUT_TOOL_NOT_EXECUTED = 'Output tool not used - a final result was already processed.'
+"""Default message when an output tool call is skipped because a result was already found."""
+
+DEFAULT_OUTPUT_VALIDATION_FAILED = 'Output tool not used - output failed validation.'
+"""Default message when an output tool fails validation."""
+
+DEFAULT_FUNCTION_TOOL_NOT_EXECUTED = 'Tool not executed - a final result was already processed.'
+"""Default message when a function tool call is skipped because a result was already found."""
+
+DEFAULT_TOOL_CALL_DENIED = 'The tool call was denied.'
+"""Default message when a tool call is denied by an approval handler."""
+
+DEFAULT_MODEL_RETRY = 'Fix the errors and try again.'
+"""Default message appended to retry prompts."""
+
+DEFAULT_PROMPTED_OUTPUT_TEMPLATE = dedent(
+    """
+    Always respond with a JSON object that's compatible with this schema:
+
+    {schema}
+
+    Don't include any text or Markdown fencing before or after.
+    """
+)
+"""Default template for prompted output schema instructions."""
+
+
+def default_validation_feedback(content: str | list[pydantic_core.ErrorDetails]) -> str:
```

Comment:
We can't require `str` here?

Reply:
Isn't it used only for the case where the content is a string, I added this assert to satisfy the type checker otherwise we would need two templates for when content is a string or list of errors? Or do you mean something else altogether?